pkg/storage/fs/ocis/upload.go

```diff
@@ -0,0 +1,499 @@
+
+// WriteChunk writes the stream from the reader to the given offset of the upload
+func (upload *fileUpload) WriteChunk(ctx context.Context, offset int64, src io.Reader) (int64, error) {
+	file, err := os.OpenFile(upload.binPath, os.O_WRONLY|os.O_APPEND, defaultFilePerm)
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	n, err := io.Copy(file, src)
+
+	// If the HTTP PATCH request gets interrupted in the middle (e.g. because
+	// the user wants to pause the upload), Go's net/http returns an io.ErrUnexpectedEOF.
+	// However, for the ocis driver it's not important whether the stream has ended
+	// on purpose or accidentally.
+	if err != nil {
+		if err != io.ErrUnexpectedEOF {
+			return n, err
+		}
+	}
+
+	upload.info.Offset += n
+	err = upload.writeInfo() // TODO info is written here ... we need to truncate in DiscardChunk
+
+	return n, err
+}
+
+// GetReader returns an io.Reader for the upload
+func (upload *fileUpload) GetReader(ctx context.Context) (io.Reader, error) {
+	return os.Open(upload.binPath)
+}
+
+// writeInfo updates the entire information. Everything will be overwritten.
+func (upload *fileUpload) writeInfo() error {
+	data, err := json.Marshal(upload.info)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(upload.infoPath, data, defaultFilePerm)
+}
+
+// FinishUpload finishes an upload and moves the file to the internal destination
+func (upload *fileUpload) FinishUpload(ctx context.Context) error {
+
+	n := &NodeInfo{
+		pw:       upload.fs.pw,
+		ID:       upload.info.Storage["NodeId"],
+		ParentID: upload.info.Storage["NodeParentId"],
+		Name:     upload.info.Storage["NodeName"],
+	}
+
+	if n.ID == "" {
+		n.ID = uuid.New().String()
+	}
+	targetPath := filepath.Join(upload.fs.conf.Root, "nodes", n.ID)
+
+	// if target exists create new version
+	if fi, err := os.Stat(targetPath); err == nil {
+		// versions are stored alongside the actual file, so a rename can be efficient and does not cross storage / partition boundaries
+		versionsPath := filepath.Join(upload.fs.conf.Root, "nodes", n.ID+".REV."+fi.ModTime().UTC().Format(time.RFC3339Nano))
+
+		if err := os.Rename(targetPath, versionsPath); err != nil {
+			log := appctx.GetLogger(upload.ctx)
+			log.Err(err).Interface("info", upload.info).
+				Str("binPath", upload.binPath).
+				Str("targetPath", targetPath).
+				Msg("ocisfs: could not create version")
+			return err
+		}
+	}
+
+	// now rename the upload to the target path
+	// TODO put uploads on the same underlying storage as the destination dir?
+	// TODO trigger a workflow as the final rename might eg involve antivirus scanning
+	if err := os.Rename(upload.binPath, targetPath); err != nil {
+		log := appctx.GetLogger(upload.ctx)
+		log.Err(err).Interface("info", upload.info).
+			Str("binPath", upload.binPath).
+			Str("targetPath", targetPath).
+			Msg("ocisfs: could not rename")
+		return err
+	}
+
+	if err := xattr.Set(targetPath, "user.ocis.parentid", []byte(n.ParentID)); err != nil {
```

Comment:
wondering if we're going too low-level here and whether the upload code should talk to a higher level API, one that simply configures the node correctly, with xattrs and children.